File: cipherforge-mobile/ios/Runner/IntegrityCheck.swift
// =============================================================================
// IntegrityCheck.swift — iOS Native Security Checks
// =============================================================================
//
// This Swift file provides native iOS security checks that complement the
// Dart-side environment integrity service:
//
// 1. **Screen Capture Detection** — Detects screen recording and mirroring.
// 2. **Frida Detection** — Checks for the Frida instrumentation framework.
// 3. **Jailbreak Detection** — Additional jailbreak checks beyond the
//    flutter_jailbreak_detection plugin.
//
// INTEGRATION:
// - This is registered as a FlutterMethodChannel handler in the AppDelegate.
// - The Dart side calls these methods via:
//   MethodChannel('com.securevault/runtime_shield')
//
// IMPORTANT: This is a placeholder demonstrating detection patterns.
// In a production app, you would:
// - Use Apple's App Attest API for server-side device integrity verification.
// - Implement checks in C/C++ for harder reverse engineering.
// - Obfuscate string constants used in detection.
// - Combine multiple detection methods for defense in depth.
//
// LIMITATIONS:
// - Sophisticated jailbreaks (e.g., checkra1n with rootless) can bypass
//   most detection methods.
// - Frida can patch its own detection before these checks execute.
// - These are deterrents, not guarantees.
// =============================================================================

import UIKit
import Flutter

// =============================================================================
// APP DELEGATE EXTENSION
// =============================================================================

/// Extension on AppDelegate to register the security MethodChannel.
///
/// In your AppDelegate.swift, add this call in didFinishLaunchingWithOptions:
///
/// ```swift
/// IntegrityCheckPlugin.register(with: controller)
/// ```
class IntegrityCheckPlugin {

    /// The MethodChannel name — must match the Dart side exactly.
    private static let channelName = "com.securevault/runtime_shield"

    /// Registers the MethodChannel handler with the Flutter engine.
    ///
    /// - Parameter controller: The root FlutterViewController.
    static func register(with controller: FlutterViewController) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: controller.binaryMessenger
        )

        channel.setMethodCallHandler { (call, result) in
            handleMethodCall(call: call, result: result, controller: controller)
        }
    }

    /// Routes MethodChannel calls to the appropriate handler.
    private static func handleMethodCall(
        call: FlutterMethodCall,
        result: @escaping FlutterResult,
        controller: FlutterViewController
    ) {
        switch call.method {
        case "enableScreenshotProtection":
            enableScreenshotProtection(controller: controller)
            result(true)

        case "disableScreenshotProtection":
            disableScreenshotProtection(controller: controller)
            result(true)

        case "checkIntegrity":
            let isClean = performIntegrityChecks()
            result(isClean)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // =========================================================================
    // SCREENSHOT PROTECTION
    // =========================================================================

    /// Enables screenshot protection on iOS.
    ///
    /// Unlike Android's FLAG_SECURE, iOS does not have a direct equivalent.
    /// Instead, we use a combination of techniques:
    ///
    /// 1. **UITextField trick**: Adding a secure UITextField to the window
    ///    hierarchy can prevent screenshots (the content appears blank).
    ///    This exploits the same mechanism that prevents the password
    ///    dot characters from appearing in screenshots.
    ///
    /// 2. **Screen capture observation**: We observe UIScreen.isCaptured
    ///    to detect when screen recording or mirroring is active, and
    ///    overlay the content with a privacy screen.
    ///
    /// NOTE: iOS 17+ has improved screenshot APIs. The effectiveness of
    /// these techniques may vary across iOS versions.
    private static func enableScreenshotProtection(controller: FlutterViewController) {
        DispatchQueue.main.async {
            // Technique 1: Observe screen capture state.
            // When isCaptured becomes true, overlay the view.
            NotificationCenter.default.addObserver(
                forName: UIScreen.capturedDidChangeNotification,
                object: nil,
                queue: .main
            ) { _ in
                if UIScreen.main.isCaptured {
                    // Screen is being recorded/mirrored — show privacy overlay.
                    showPrivacyOverlay(on: controller)
                } else {
                    // Recording stopped — remove overlay.
                    hidePrivacyOverlay(on: controller)
                }
            }

            // Check current state in case recording is already active.
            if UIScreen.main.isCaptured {
                showPrivacyOverlay(on: controller)
            }

            // Technique 2: Add a secure text field to prevent screenshots.
            // This is a well-known trick — iOS blanks out secure text fields
            // in screenshots, and we can use this to protect the entire window.
            let secureField = UITextField()
            secureField.isSecureTextEntry = true
            secureField.isUserInteractionEnabled = false

            // Add it behind all other content.
            if let window = controller.view.window {
                window.addSubview(secureField)
                secureField.centerYAnchor.constraint(equalTo: window.centerYAnchor).isActive = true
                secureField.centerXAnchor.constraint(equalTo: window.centerXAnchor).isActive = true

                // Make the text field's layer the window's layer mask.
                // This causes the entire window to be treated as "secure content"
                // by the screenshot mechanism.
                window.layer.superlayer?.addSublayer(secureField.layer)
                secureField.layer.sublayers?.first?.addSublayer(window.layer)
            }
        }
    }

    /// Disables screenshot protection (for testing only).
    private static func disableScreenshotProtection(controller: FlutterViewController) {
        DispatchQueue.main.async {
            NotificationCenter.default.removeObserver(
                controller,
                name: UIScreen.capturedDidChangeNotification,
                object: nil
            )
            hidePrivacyOverlay(on: controller)
        }
    }

    /// Privacy overlay tag for identification.
    private static let privacyOverlayTag = 999

    /// Shows a privacy overlay when screen capture is detected.
    ///
    /// This covers the entire window with a blur effect and a message,
    /// preventing the recorded content from containing sensitive data.
    private static func showPrivacyOverlay(on controller: FlutterViewController) {
        guard controller.view.viewWithTag(privacyOverlayTag) == nil else { return }

        let overlay = UIView(frame: controller.view.bounds)
        overlay.tag = privacyOverlayTag
        overlay.backgroundColor = UIColor(red: 0.05, green: 0.07, blue: 0.09, alpha: 1.0)
        overlay.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        let blurEffect = UIBlurEffect(style: .dark)
        let blurView = UIVisualEffectView(effect: blurEffect)
        blurView.frame = overlay.bounds
        blurView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        overlay.addSubview(blurView)

        let label = UILabel()
        label.text = "🔒 Content hidden for security"
        label.textColor = .white
        label.textAlignment = .center
        label.font = UIFont.systemFont(ofSize: 18, weight: .semibold)
        label.translatesAutoresizingMaskIntoConstraints = false
        overlay.addSubview(label)

        NSLayoutConstraint.activate([
            label.centerXAnchor.constraint(equalTo: overlay.centerXAnchor),
            label.centerYAnchor.constraint(equalTo: overlay.centerYAnchor)
        ])

        controller.view.addSubview(overlay)
    }

    /// Removes the privacy overlay when screen capture stops.
    private static func hidePrivacyOverlay(on controller: FlutterViewController) {
        controller.view.viewWithTag(privacyOverlayTag)?.removeFromSuperview()
    }

    // =========================================================================
    // INTEGRITY CHECKS
    // =========================================================================

    /// Performs all native integrity checks.
    ///
    /// Returns true if the environment appears clean.
    /// Returns false if any threat is detected.
    private static func performIntegrityChecks() -> Bool {
        let isFridaDetected = checkForFrida()
        let isJailbroken = checkForJailbreak()

        return !isFridaDetected && !isJailbroken
    }

    // =========================================================================
    // FRIDA DETECTION
    // =========================================================================

    /// Checks for the Frida instrumentation framework on iOS.
    ///
    /// Frida on iOS works by:
    /// 1. Injecting FridaGadget.dylib into the app process.
    /// 2. Or running frida-server on the device (requires jailbreak).
    ///
    /// Detection methods:
    /// 1. Check for Frida's default port (27042).
    /// 2. Scan loaded dylibs for Frida-related names.
    /// 3. Check for the FridaGadget environment variable.
    private static func checkForFrida() -> Bool {
        // Check 1: Frida default port
        if isFridaPortOpen() { return true }

        // Check 2: Frida dylibs in loaded images
        if isFridaDylibLoaded() { return true }

        // Check 3: Frida environment variables
        if isFridaEnvSet() { return true }

        return false
    }

    /// Checks if Frida's default port (27042) is listening.
    private static func isFridaPortOpen() -> Bool {
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = UInt16(27042).bigEndian
        addr.sin_addr.s_addr = inet_addr("127.0.0.1")

        let sock = socket(AF_INET, SOCK_STREAM, 0)
        guard sock >= 0 else { return false }

        let result = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                connect(sock, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }

        close(sock)
        return result == 0 // 0 = connection succeeded = port is open
    }

    /// Scans loaded dynamic libraries for Frida-related names.
    ///
    /// When Frida injects into an iOS process, it loads dylibs with
    /// recognizable names like "FridaGadget" or "frida-agent".
    private static func isFridaDylibLoaded() -> Bool {
        let imageCount = _dyld_image_count()

        for i in 0..<imageCount {
            guard let imageName = _dyld_get_image_name(i) else { continue }
            let name = String(cString: imageName).lowercased()

            if name.contains("frida") || name.contains("gadget") {
                return true
            }
        }

        return false
    }

    /// Checks for Frida-related environment variables.
    private static func isFridaEnvSet() -> Bool {
        let fridaEnvVars = [
            "FRIDA_AGENT_PATH",
            "FRIDA_GADGET_CONFIG"
        ]

        for envVar in fridaEnvVars {
            if ProcessInfo.processInfo.environment[envVar] != nil {
                return true
            }
        }

        return false
    }

    // =========================================================================
    // JAILBREAK DETECTION (supplementary)
    // =========================================================================

    /// Additional jailbreak detection checks.
    ///
    /// These supplement flutter_jailbreak_detection with native checks:
    /// 1. Check for common jailbreak files and directories.
    /// 2. Check for unauthorized URL schemes (Cydia, Sileo).
    /// 3. Check if the app can write outside its sandbox.
    /// 4. Check for symbolic links in system directories.
    ///
    /// IMPORTANT: These checks should be performed from native code
    /// because Dart-side checks can be more easily bypassed by hooking
    /// the method channel.
    private static func checkForJailbreak() -> Bool {
        // Check 1: Jailbreak-related files
        let jailbreakPaths = [
            "/Applications/Cydia.app",
            "/Applications/Sileo.app",
            "/Library/MobileSubstrate/MobileSubstrate.dylib",
            "/bin/bash",
            "/usr/sbin/sshd",
            "/etc/apt",
            "/usr/bin/ssh",
            "/private/var/lib/apt",
            "/private/var/lib/cydia",
            "/private/var/stash",
            "/var/cache/apt",
            "/var/lib/cydia",
            "/var/tmp/cydia.log",
            "/usr/libexec/cydia",
        ]

        for path in jailbreakPaths {
            if FileManager.default.fileExists(atPath: path) {
                return true
            }
        }

        // Check 2: Cydia URL scheme
        if let url = URL(string: "cydia://package/com.example.package") {
            if UIApplication.shared.canOpenURL(url) {
                return true
            }
        }

        // Check 3: Write outside sandbox
        // On a non-jailbroken device, the app cannot write to /private.
        let testPath = "/private/jailbreak_test_\(UUID().uuidString)"
        do {
            try "test".write(toFile: testPath, atomically: true, encoding: .utf8)
            // If we got here, writing succeeded — device is jailbroken.
            try? FileManager.default.removeItem(atPath: testPath)
            return true
        } catch {
            // Write failed — expected on non-jailbroken devices.
        }

        // Check 4: Fork detection
        // Non-jailbroken apps cannot fork processes.
        // Calling fork() returns -1 on a sandboxed app.
        #if !targetEnvironment(simulator)
        let forkResult = fork()
        if forkResult >= 0 {
            // Fork succeeded — device is jailbroken.
            if forkResult > 0 {
                // Parent process — kill the child.
                kill(forkResult, SIGTERM)
            }
            return true
        }
        #endif

        return false
    }
}

File: cipherforge-mobile/android/app/src/main/kotlin/com/securevault/IntegrityCheck.kt
// =============================================================================
// IntegrityCheck.kt — Android Native Security Checks
// =============================================================================
//
// This Kotlin file provides native Android security checks that cannot be
// performed effectively from the Dart side:
//
// 1. **FLAG_SECURE** — Prevents screenshots and screen recording.
// 2. **Frida Detection** — Checks for the Frida instrumentation framework.
// 3. **Xposed Detection** — Checks for the Xposed hooking framework.
// 4. **Root Detection** — Additional root checks beyond flutter_jailbreak_detection.
//
// INTEGRATION:
// - This is registered as a MethodChannel handler in the Flutter activity.
// - The Dart side calls these methods via MethodChannel('com.securevault/runtime_shield').
//
// IMPORTANT: This is a placeholder that demonstrates the detection patterns.
// In a production app, you would:
// - Obfuscate this code with ProGuard/R8 to prevent easy bypass.
// - Implement the checks in native C/C++ via JNI for harder reverse engineering.
// - Use Google Play Integrity API for server-side attestation.
// - Combine multiple detection methods for defense in depth.
//
// LIMITATIONS:
// - Magisk Hide / DenyList can bypass most root detection.
// - Frida can unhook its own detection before it's detected.
// - These are speed bumps, not absolute barriers.
// =============================================================================

package com.securevault

import android.os.Bundle
import android.view.WindowManager
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.net.Socket

/// Main Flutter Activity with integrated security checks.
///
/// Extends FlutterActivity to:
/// - Register the MethodChannel for Dart-to-native communication.
/// - Handle FLAG_SECURE for screenshot protection.
/// - Provide native integrity checking methods.
class MainActivity : FlutterActivity() {

    /// The MethodChannel name — must match the Dart side exactly.
    private val CHANNEL = "com.securevault/runtime_shield"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Enable FLAG_SECURE by default at activity creation.
        // This prevents screenshots even before Flutter renders the first frame.
        window.setFlags(
            WindowManager.LayoutParams.FLAG_SECURE,
            WindowManager.LayoutParams.FLAG_SECURE
        )
    }

    override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
        super.configureFlutterEngine(flutterEngine)

        // Register the MethodChannel handler.
        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL)
            .setMethodCallHandler { call, result ->
                handleMethodCall(call, result)
            }
    }

    /// Routes MethodChannel calls to the appropriate handler.
    private fun handleMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "enableScreenshotProtection" -> {
                enableScreenshotProtection()
                result.success(true)
            }
            "disableScreenshotProtection" -> {
                disableScreenshotProtection()
                result.success(true)
            }
            "checkIntegrity" -> {
                val isClean = performIntegrityChecks()
                result.success(isClean)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // =========================================================================
    // FLAG_SECURE — Screenshot Protection
    // =========================================================================

    /// Enables FLAG_SECURE on the activity window.
    ///
    /// When FLAG_SECURE is set:
    /// - System screenshots are blocked (returns a black/blank image).
    /// - MediaProjection (screen recording) shows a black window.
    /// - The recent apps switcher shows a blank preview.
    /// - Cast/Miracast does not show the app content.
    ///
    /// This is the most effective screenshot protection available on Android.
    private fun enableScreenshotProtection() {
        runOnUiThread {
            window.setFlags(
                WindowManager.LayoutParams.FLAG_SECURE,
                WindowManager.LayoutParams.FLAG_SECURE
            )
        }
    }

    /// Disables FLAG_SECURE (for testing only — not recommended in production).
    private fun disableScreenshotProtection() {
        runOnUiThread {
            window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
        }
    }

    // =========================================================================
    // INTEGRITY CHECKS
    // =========================================================================

    /// Performs all native integrity checks.
    ///
    /// Returns true if the environment appears clean (no threats detected).
    /// Returns false if any check fails (potential compromise).
    private fun performIntegrityChecks(): Boolean {
        val isFridaDetected = checkForFrida()
        val isXposedDetected = checkForXposed()
        val isRooted = checkForRoot()

        // Return true only if ALL checks pass (no threats).
        return !isFridaDetected && !isXposedDetected && !isRooted
    }

    // =========================================================================
    // FRIDA DETECTION
    // =========================================================================

    /// Checks for the Frida instrumentation framework.
    ///
    /// Frida is a powerful dynamic instrumentation toolkit that can:
    /// - Hook any function at runtime.
    /// - Read/write process memory.
    /// - Intercept network traffic.
    /// - Bypass all security checks.
    ///
    /// Detection methods:
    /// 1. Check if Frida's default port (27042) is listening.
    /// 2. Scan /proc/self/maps for frida-agent shared libraries.
    /// 3. Check for frida-server binary in common locations.
    ///
    /// LIMITATIONS:
    /// - Frida can be configured to use a non-default port.
    /// - Frida can be injected without frida-server (via ptrace or root).
    /// - Sophisticated attackers can patch these checks before they run.
    private fun checkForFrida(): Boolean {
        // Check 1: Frida default port
        if (isFridaPortOpen()) return true

        // Check 2: Frida agent in memory maps
        if (isFridaAgentLoaded()) return true

        // Check 3: Frida server binary
        if (isFridaServerPresent()) return true

        return false
    }

    /// Checks if Frida's default port (27042) is listening.
    ///
    /// Frida-server listens on port 27042 by default for incoming connections
    /// from the Frida client on the host machine.
    private fun isFridaPortOpen(): Boolean {
        return try {
            val socket = Socket("127.0.0.1", 27042)
            socket.close()
            true // Port is open — Frida likely running
        } catch (e: Exception) {
            false // Port closed — no Frida on default port
        }
    }

    /// Scans /proc/self/maps for Frida agent shared libraries.
    ///
    /// When Frida injects its agent into a process, it loads a shared library
    /// (typically named frida-agent-*.so) into the process's memory space.
    /// This appears in /proc/self/maps.
    private fun isFridaAgentLoaded(): Boolean {
        return try {
            val maps = File("/proc/self/maps").readText()
            maps.contains("frida") || maps.contains("gadget")
        } catch (e: Exception) {
            false // Can't read maps — assume clean
        }
    }

    /// Checks for frida-server binary in common locations.
    private fun isFridaServerPresent(): Boolean {
        val fridaPaths = listOf(
            "/data/local/tmp/frida-server",
            "/data/local/tmp/re.frida.server",
            "/sdcard/frida-server"
        )

        return fridaPaths.any { File(it).exists() }
    }

    // =========================================================================
    // XPOSED DETECTION
    // =========================================================================

    /// Checks for the Xposed Framework.
    ///
    /// Xposed allows modifying the behavior of any Android app without
    /// changing its APK. It works by:
    /// 1. Replacing /system/bin/app_process with a modified version.
    /// 2. Loading Xposed modules that hook into app methods.
    ///
    /// For a password manager, Xposed could:
    /// - Hook the decryption methods to capture plaintext passwords.
    /// - Hook the clipboard to intercept copied passwords.
    /// - Hook the biometric callback to bypass authentication.
    ///
    /// Detection methods:
    /// 1. Check for Xposed Installer package.
    /// 2. Check for the XposedBridge class in the class path.
    /// 3. Check for Xposed-related files in /system.
    private fun checkForXposed(): Boolean {
        // Check 1: Xposed Installer package
        val xposedPackages = listOf(
            "de.robv.android.xposed.installer",
            "org.meowcat.edxposed.manager",
            "org.lsposed.manager"
        )

        for (pkg in xposedPackages) {
            try {
                packageManager.getPackageInfo(pkg, 0)
                return true // Xposed manager found
            } catch (e: Exception) {
                // Package not found — continue checking
            }
        }

        // Check 2: XposedBridge class
        try {
            Class.forName("de.robv.android.xposed.XposedBridge")
            return true // Xposed framework loaded
        } catch (e: ClassNotFoundException) {
            // Not found — continue
        }

        // Check 3: Xposed-related system files
        val xposedFiles = listOf(
            "/system/framework/XposedBridge.jar",
            "/system/lib/libxposed_art.so",
            "/system/lib64/libxposed_art.so"
        )

        return xposedFiles.any { File(it).exists() }
    }

    // =========================================================================
    // ROOT DETECTION (supplementary)
    // =========================================================================

    /// Additional root detection checks beyond flutter_jailbreak_detection.
    ///
    /// These checks supplement the Dart-side detection with native checks
    /// that are harder to bypass from the Dart level.
    ///
    /// Detection methods:
    /// 1. Check for su binary in common locations.
    /// 2. Check for Magisk-specific files.
    /// 3. Check build tags for "test-keys" (custom ROM indicator).
    private fun checkForRoot(): Boolean {
        // Check 1: su binary
        val suPaths = listOf(
            "/system/bin/su",
            "/system/xbin/su",
            "/sbin/su",
            "/data/local/bin/su",
            "/data/local/xbin/su"
        )

        if (suPaths.any { File(it).exists() }) return true

        // Check 2: Magisk files
        val magiskPaths = listOf(
            "/sbin/.magisk",
            "/data/adb/magisk",
            "/data/adb/magisk.db"
        )

        if (magiskPaths.any { File(it).exists() }) return true

        // Check 3: Build tags
        val buildTags = android.os.Build.TAGS
        if (buildTags != null && buildTags.contains("test-keys")) return true

        // Check 4: Try to execute su
        return try {
            val process = Runtime.getRuntime().exec(arrayOf("which", "su"))
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val result = reader.readLine()
            result != null && result.isNotEmpty()
        } catch (e: Exception) {
            false
        }
    }
}
